## M280 Homework4
* Author: Shuang Gao
* Date: 2018/05/31

### Question 1
For a multivariate count vector $\mathbf{x}=(x_1,\ldots,x_d)$ with batch size $|\mathbf{x}|=\sum_{j=1}^d x_j$, show that the probability mass function for Dirichlet-multinomial distribution is
$$
    f(\mathbf{x} \mid \alpha) 
	= \int_{\Delta_d} \binom{|\mathbf{x}|}{\mathbf{x}} \prod_{j=1}^d p_j^{x_j} \pi(\mathbf{p}) \, d \mathbf{p}  
    = \binom{|\mathbf{x}|}{\mathbf{x}} \frac{\prod_{j=1}^d \Gamma(\alpha_j+x_j)}{\prod_{j=1}^d \Gamma(\alpha_j)} \frac{\Gamma(|\alpha|)}{\Gamma(|\alpha|+|\mathbf{x}|)}
$$
where $\Delta_d$ is the unit simplex in $d$ dimensions and $|\alpha| = \sum_{j=1}^d \alpha_j$.

**Solution:**
$$
\begin{align}
f(\mathbf{x} \mid \alpha) 
&= \int_{\Delta_d} \binom{|\mathbf{x}|}{\mathbf{x}} \prod_{j=1}^d p_j^{x_j} \pi(\mathbf{p}) \, d \mathbf{p} \\
&= \int_{\Delta_d} \binom{|\mathbf{x}|}{\mathbf{x}} \prod_{j=1}^d p_j^{x_j}\frac{\Gamma(|\alpha|)}{\prod_{j=1}^d \Gamma(\alpha_j)} \prod_{j=1}^d p_j^{\alpha_j-1}\,d \mathbf{p}\\
&= \binom{|\mathbf{x}|}{\mathbf{x}} 
\frac{\Gamma(|\alpha|)}{\prod_{j=1}^d \Gamma(\alpha_j)} \int_{\Delta_d} \prod_{j=1}^d p_j^{x_j+\alpha_j-1}\, d\mathbf{p}
\end{align}
$$

Because $\mathbf{p} = (p_1,\ldots, p_d)$ follow a Dirichlet distribution with parameter vector $\alpha = (\alpha_1,\ldots, \alpha_d)$, $\alpha_j>0$, and density
$$
\begin{eqnarray*}
	\pi(\mathbf{p}) =  \frac{\Gamma(|\alpha|)}{\prod_{j=1}^d \Gamma(\alpha_j)} \prod_{j=1}^d p_j^{\alpha_j-1},
\end{eqnarray*} 
$$
where $|\alpha|=\sum_{j=1}^d \alpha_j$.
We can get,
$$
\begin{align}
\int \pi(\mathbf{p})\,d\mathbf{p} 
&= \int \frac{\Gamma(|\alpha|)}{\prod_{j=1}^d \Gamma(\alpha_j)} \prod_{j=1}^d p_j^{\alpha_j-1} \,d\mathbf{p}=1\\
& \therefore \int \prod_{j=1}^d p_j^{\alpha_j-1}\,d\mathbf{p}=\frac{\prod_{j=1}^d\Gamma{(\alpha_j})}{\Gamma{(|\alpha|)}}\hspace{12mm}(1)
\end{align}
$$
In order to get the derivation of $f(\mathbf{x} \mid \alpha) $, Plug $\alpha_j=\alpha_j+x_j$ into equation (1).
Then we have,
$$
\begin{align}
\int \prod_{j=1}^d p_j^{\alpha_j+x_j-1}\,d\mathbf{p}=\frac{\prod_{j=1}^d\Gamma{(\alpha_j}+x_j)}{\Gamma{(|\alpha|+|\mathbf{x}|)}}
\end{align}
$$
Then
$$
\begin{align}
f(\mathbf{x} \mid \alpha) 
&= \binom{|\mathbf{x}|}{\mathbf{x}} 
\frac{\Gamma(|\alpha|)}{\prod_{j=1}^d \Gamma(\alpha_j)} \frac{\prod_{j=1}^d\Gamma{(\alpha_j}+x_j)}{\Gamma{(|\alpha|+|\mathbf{x}|)}}\\
&=  \binom{|\mathbf{x}|}{\mathbf{x}} \frac{\prod_{j=1}^d \Gamma(\alpha_j+x_j)}{\prod_{j=1}^d \Gamma(\alpha_j)} \frac{\Gamma(|\alpha|)}{\Gamma(|\alpha|+|\mathbf{x}|)}
\end{align}
$$


### Question 2

Given independent data points $\mathbf{x}_1, \ldots, \mathbf{x}_n$, show that the log-likelihood is
$$
L(\alpha) = \sum_{i=1}^n \ln \binom{|\mathbf{x}_i|}{\mathbf{x}_i} + \sum_{i=1}^n \sum_{j=1}^d [\ln \Gamma(\alpha_j + x_{ij}) - \ln \Gamma(\alpha_j)] - \sum_{i=1}^n [\ln \Gamma(|\alpha|+|\mathbf{x}_i|) - \ln \Gamma(|\alpha|)]
$$
Is the log-likelihood a concave function?

**Solution:**
$$
\begin{align}
& \because \mathbf{x_1},...,\mathbf{x_n} \text{ are independent}\\ 
& \therefore f(\mathbf{x_1},...,\mathbf{x_n} \mid \alpha) 
= f(\mathbf{x_1}\mid \alpha)...f(\mathbf{x_n\mid \alpha})
\end{align}
$$
$$
\begin{align}
\therefore f(\mathbf{x_1},...,\mathbf{x_n} \mid \alpha) 
= \prod_{i=1}^n \binom{|\mathbf{x_i}|}{\mathbf{x_i}} \frac{\prod_{j=1}^d \Gamma(\alpha_j+x_{ij})}{\prod_{j=1}^d \Gamma(\alpha_j)} \frac{\Gamma(|\alpha|)}{\Gamma(|\alpha|+|\mathbf{x_i}|)}
\end{align}
$$
$$
\begin{align}
\therefore L(\alpha) 
&= \ln (\prod_{i=1}^n \binom{|\mathbf{x_i}|}{\mathbf{x_i}} \frac{\prod_{j=1}^d \Gamma(\alpha_j+x_{ij})}{\prod_{j=1}^d \Gamma(\alpha_j)} \frac{\Gamma(|\alpha|)}{\Gamma(|\alpha|+|\mathbf{x_i}|)})\\
&= \sum_{i=1}^n \ln \binom{|\mathbf{x}_i|}{\mathbf{x}_i} + \sum_{i=1}^n \sum_{j=1}^d [\ln \Gamma(\alpha_j + x_{ij}) - \ln \Gamma(\alpha_j)] - \sum_{i=1}^n [\ln \Gamma(|\alpha|+|\mathbf{x}_i|) - \ln \Gamma(|\alpha|)]
\end{align}
$$

* The function is not concave.

### Question 3

Write Julia function to compute the log-density of the Dirichlet-multinomial distribution.

The log density of the Dirichlet-multinomial distribution for one data point $\mathbf{x}$ is shown as follow:
$$
\begin{align}
L(\alpha) 
&= \ln \binom{|\mathbf{x}|}{\mathbf{x}} +  \sum_{j=1}^d [\ln \Gamma(\alpha_j + x_{j}) - \ln \Gamma(\alpha_j)] - [\ln \Gamma(|\alpha|+|\mathbf{x}|) - \ln \Gamma(|\alpha|)]\\
&=  \ln |\mathbf{x}|! - \ln(x_1!)-...-\ln(x_{64}!)+ \sum_{j=1}^{64} [\ln \Gamma(\alpha_j + x_{j}) - \ln \Gamma(\alpha_j)] - [\ln \Gamma(|\alpha|+|\mathbf{x}|) - \ln \Gamma(|\alpha|)]\\
&=  \ln |\mathbf{x}|! - [\ln \Gamma(|\alpha|+|\mathbf{x}|) - \ln \Gamma(|\alpha|)]+ \sum_{j=1}^{64} [\ln \Gamma(\alpha_j + x_{j}) - \ln \Gamma(\alpha_j)-\ln(x_j!)]
\end{align}
$$



In [14]:
"""
    dirmult_logpdf(x::Vector, α::Vector)
    
Compute the log-pdf of Dirichlet-multinomial distribution with parameter `α` 
at data point `x`.
"""
function dirmult_logpdf(x::Vector, α::Vector)
    # store the sum of x and α
    sumx = sum(x)
    sumα = sum(α)
    
    # store length of x
    lengthx = size(x, 1) 
    
    # part without summation
    logden = lfact(sumx) - lgamma(sumα + sumx) + lgamma(sumα)
    
    # part within summation
    for j in lengthx
       logden += lgamma(α[j] + x[j]) - lgamma(α[j]) - lfact(x[j])  
    end
    
    # return log density
    return logden
end

function dirmult_logpdf!(r::Vector, X::Matrix, α::Vector)
    for j in 1:size(X, 2)
        r[j] = dirmult_logpdf(X[:, j], α)
    end
    return r
end

"""
    dirmult_logpdf(X, α)
    
Compute the log-pdf of Dirichlet-multinomial distribution with parameter `α` 
at each data point in `X`. Each column of `X` is one data point.
"""
function dirmult_logpdf(X::Matrix, α::Vector)
    r = zeros(size(X, 2))
    dirmult_logpdf!(r, X, α)
end


dirmult_logpdf

### Question 4

Read in `optdigits.tra`, the training set of 3823 handwritten digits. Each row contains the 64 counts of a digit and the last element (65th element) indicates what digit it is. For grading purpose, evaluate the total log-likelihood of this data at parameter values $\alpha=(1,\ldots,1)$ using your function in Q3.

In [43]:
# read in train data
path1_tra = "http://hua-zhou.github.io/teaching/biostatm280"
path2_tra = "-2018spring/hw/hw4/optdigits.tra"

train = readdlm(download(path1_tra * path2_tra), ',')

# separte last column which store the digit 
train_x = round.(Int64, train[:, 1:64])
train_digit = round.(Int64, train[:, 65]);


In [45]:
# store size of train data
length_x = size(train_x, 2)
length_obs = size(train_x, 1)

# input α vector
α = round.(Int64, ones(length_x))

sum(dirmult_logpdf(train_x', α))

-638817.993292528

* The total log-likelihood of this data at parameter values $\alpha=(1,\ldots,1)$ is around -638818.

### Question5 
Derive the score function $\nabla L(\alpha)$, observed information matrix $-d^2L(\alpha)$, and Fisher information matrix $\mathbf{E}[-d^2L(\alpha)]$ for the Dirichlet-multinomial distribution.

Comment on why Fisher scoring method is inefficient for computing MLE in this example.